packages/session-replay-browser/src/events/events-idb-store.ts

```diff
@@ -0,0 +1,308 @@
+  Events,
+  SendingSequencesIDBInput,
+  SendingSequencesIDBReturn,
+} from '../typings/session-replay';
+import { IDBStore, IDBStoreSession, RecordingStatus } from './legacy-idb-types';
+
+export const currentSequenceKey = 'sessionCurrentSequence';
+export const sequencesToSendKey = 'sequencesToSend';
+export const remoteConfigKey = 'remoteConfig';
+
+export interface SessionReplayDB extends DBSchema {
+  sessionCurrentSequence: {
+    key: number;
+    value: {
+      sessionId: number;
+      events: Events;
+    };
+  };
+  sequencesToSend: {
+    key: number;
+    value: SendingSequencesIDBInput;
+    indexes: { sessionId: number };
+  };
+}
+
+export const keyValDatabaseExists = function (): Promise<IDBDatabase | void> {
+  const globalScope = getGlobalScope();
+  return new Promise((resolve) => {
+    if (globalScope) {
+      const request = globalScope.indexedDB.open('keyval-store');
+      request.onupgradeneeded = function () {
+        if (request.result.version === 1) {
+          request.result.close();
+          request.transaction && request.transaction.abort();
+          globalScope.indexedDB.deleteDatabase('keyval-store');
+          resolve();
+        }
+      };
+      request.onsuccess = function () {
+        resolve(request.result);
+      };
+    }
+  });
+};
+
+export const defineObjectStores = (db: IDBPDatabase<SessionReplayDB>) => {
+  let sequencesStore;
+  let currentSequenceStore;
+  if (!db.objectStoreNames.contains(currentSequenceKey)) {
+    currentSequenceStore = db.createObjectStore(currentSequenceKey, {
+      keyPath: 'sessionId',
+    });
+  }
+  if (!db.objectStoreNames.contains(sequencesToSendKey)) {
+    sequencesStore = db.createObjectStore(sequencesToSendKey, {
+      keyPath: 'sequenceId',
+      autoIncrement: true,
+    });
+    sequencesStore.createIndex('sessionId', 'sessionId');
+  }
+  return {
+    sequencesStore,
+    currentSequenceStore,
+  };
+};
+
+export const createStore = async (dbName: string) => {
+  return await openDB<SessionReplayDB>(dbName, 1, {
+    upgrade: defineObjectStores,
+  });
+};
+
+export class SessionReplayEventsIDBStore implements AmplitudeSessionReplayEventsIDBStore {
+  apiKey: string;
+  db: IDBPDatabase<SessionReplayDB> | undefined;
+  loggerProvider: ILogger;
+  storageKey = '';
+  maxPersistedEventsSize = MAX_EVENT_LIST_SIZE_IN_BYTES;
+  interval = MIN_INTERVAL;
+  timeAtLastSplit: number | null = null;
+
+  constructor({ loggerProvider, apiKey }: { loggerProvider: ILogger; apiKey: string }) {
+    this.loggerProvider = loggerProvider;
+    this.apiKey = apiKey;
+  }
+
+  async initialize(sessionId?: number) {
+    const dbName = `${this.apiKey.substring(0, 10)}_amp_session_replay_events`;
+    this.db = await createStore(dbName);
+    this.timeAtLastSplit = Date.now(); // Initialize this so we have a point of comparison when events are recorded
+    await this.transitionFromKeyValStore(sessionId);
+  }
```

Comment:
My suggestion here is not to have a public constructor, but instead make it private and initialize is the true constructor that's async. Roughly, I'm thinking it'll look like this in the end.
```suggestion
db: IDBPDatabase<SessionReplayDB>;
  loggerProvider: ILogger;
  storageKey = '';
  maxPersistedEventsSize = MAX_EVENT_LIST_SIZE_IN_BYTES;
  interval = MIN_INTERVAL;
  timeAtLastSplit: number;

  private constructor({ loggerProvider, apiKey, db, timeAtLastSplit }: { loggerProvider: ILogger; apiKey: string; db: IDBPDatabase<SessionReplayDB>; timeAtLastSplit: number }) {
    this.loggerProvider = loggerProvider;
    this.apiKey = apiKey;
    this.db = db;
    this.timeAtLastSplit = timeAtLastSplit;
  }

  static async initialize({ loggerProvider, apiKey, sessionId }: { loggerProvider: ILogger; apiKey: string; sessionId?: number}): Promise<SessionReplayEventsIDBStore> {
    const dbName = `${apiKey.substring(0, 10)}_amp_session_replay_events`;
    const db = await createStore(dbName);
    const timeAtLastSplit = Date.now(); // Initialize this so we have a point of comparison when events are recorded
    const store = new SessionReplayEventsIDBStore({loggerProvider, apiKey, db, timeAtLastSplit})
    await store.transitionFromKeyValStore(sessionId);
    return store
  }
```

The reason I like this is that it can remove some possibly null fields and users of this class will never need to worry about the pattern of constructor then initialize.

Reply:
I thought about this a lot too (probably too much haha), and tried an implementation like you're mentioning, based on the suggestion here: https://dev.to/somedood/the-proper-way-to-write-async-constructors-in-javascript-1o8c, sounds like the same approach. I ended up backing away from it earlier because at the time I found it harder to work with/wrap my head around than the create fn I exposed at the bottom of this file, but I'm definitely not against it. I'm going to leave it as is for now, but I'm totally open to this pattern moving forward and definitely happy to consider refactoring it down the line